examples/python/auth/async_customized_auth_client.py

```diff
@@ -0,0 +1,101 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Client of the Python AsyncIO example of customizing authentication mechanism."""
+
+import argparse
+import asyncio
+import logging
+
+import grpc
+
+import _credentials
+
+helloworld_pb2, helloworld_pb2_grpc = grpc.protos_and_services(
+    "helloworld.proto")
+
+_LOGGER = logging.getLogger(__name__)
+_LOGGER.setLevel(logging.INFO)
+
+_SERVER_ADDR_TEMPLATE = 'localhost:%d'
+_SIGNATURE_HEADER_KEY = 'x-signature'
+
+
+class AuthGateway(grpc.AuthMetadataPlugin):
+
+    def __call__(self, context: grpc.AuthMetadataContext,
+                 callback: grpc.AuthMetadataPluginCallback) -> None:
+        """Implements authentication by passing metadata to a callback.
+
+        Implementations of this method must not block.
+
+        Args:
+          context: An AuthMetadataContext providing information on the RPC that
+            the plugin is being called to authenticate.
+          callback: An AuthMetadataPluginCallback to be invoked either
+            synchronously or asynchronously.
+        """
+        # Example AuthMetadataContext object:
+        # AuthMetadataContext(
+        #     service_url=u'https://localhost:50051/helloworld.Greeter',
+        #     method_name=u'SayHello')
+        signature = context.method_name[::-1]
+        callback(((_SIGNATURE_HEADER_KEY, signature),), None)
+
+
+def create_client_channel(addr: str) -> grpc.aio.Channel:
+    # Call credential object will be invoked for every single RPC
+    call_credentials = grpc.metadata_call_credentials(AuthGateway(),
+                                                      name='auth gateway')
+    # Channel credential will be valid for the entire channel
+    channel_credential = grpc.ssl_channel_credentials(
+        _credentials.ROOT_CERTIFICATE)
+    # Combining channel credentials and call credentials together
+    composite_credentials = grpc.composite_channel_credentials(
+        channel_credential,
+        call_credentials,
+    )
+    channel = grpc.aio.secure_channel(addr, composite_credentials)
+    return channel
+
+
+async def send_rpc(channel: grpc.aio.Channel) -> helloworld_pb2.HelloReply:
+    stub = helloworld_pb2_grpc.GreeterStub(channel)
+    request = helloworld_pb2.HelloRequest(name='you')
+    try:
+        response = await stub.SayHello(request)
+    except grpc.RpcError as rpc_error:
+        _LOGGER.error('Received error: %s', rpc_error)
+        return rpc_error
+    else:
+        _LOGGER.info('Received message: %s', response)
+        return response
+
+
+async def main() -> None:
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--port',
+                        nargs='?',
+                        type=int,
+                        default=50051,
+                        help='the address of server')
+    args = parser.parse_args()
+
+    channel = create_client_channel(_SERVER_ADDR_TEMPLATE % args.port)
```

Comment:
Question: Any reason not to use `async with`?

Reply:
The API to make a function usable for `async with` is [`contextlib.asynccontextmanager`](https://docs.python.org/3/library/contextlib.html#contextlib.asynccontextmanager), but only available in 3.7. Or we have to make it a class with `__aenter__` and `__aexit__`. I hope our examples can support 3.6, as our minimum supported version suggests. I can try out other styles, if you have thoughts.

Reply:
No, that's reasonable. I also saw later on in the PR that you have examples with the `async with` version. Showing off both styles is probably a good idea anyway.